proxy/Dockerfile

```diff
@@ -2,25 +2,110 @@
+COPY Cargo.toml Cargo.lock ./
+COPY proxy/Cargo.toml          ./proxy/Cargo.toml
+COPY proxy/convert             ./proxy/convert
+COPY proxy/futures-mpsc-lossy  ./proxy/futures-mpsc-lossy
+COPY proxy/grpc/Cargo.toml     ./proxy/grpc/Cargo.toml
+COPY proxy/tower-router        ./proxy/tower-router
+ARG DEBUG
+RUN if [ -n "$DEBUG" ]; \
+    then cargo build --frozen ; \
+    else cargo build --frozen --release ; \
+    fi
+#RUN if [ -n "$SKIP_TESTS" ]; \
+#    then echo "tests skipped" ; \
+#    elif [ -n "$DEBUG" ]; \
+#    then cargo test --frozen ; \
+#    else cargo test --frozen --release ; \
+#    fi
+
+## Build gRPC bindings.
 #
-# If the RELEASE arg is set and non-empty, a release artifact is built.
-FROM gcr.io/runconduit/proxy-deps:b84ed5d0 as build
+# Mocks proxy sources to ensure that dependencies, including build-time dependencies
+# needed to generate the bindings, are cached.
+FROM $RUST_IMAGE as grpc
 WORKDIR /usr/src/conduit
-# Ranked roughly from least to most likely to change. Cargo.lock is the least likely
-# because it is supposed to be cached in the deps base image.
-COPY futures-mpsc-lossy ./futures-mpsc-lossy
-COPY tower-router ./tower-router
+COPY --from=fetch $CARGO_HOME $CARGO_HOME
+COPY proxy/.empty.rs proxy/src/lib.rs
+COPY Cargo.toml Cargo.lock ./
+COPY proto                     ./proto
+COPY proxy/Cargo.toml          ./proxy/Cargo.toml
+COPY proxy/convert             ./proxy/convert
+COPY proxy/futures-mpsc-lossy  ./proxy/futures-mpsc-lossy
+COPY proxy/grpc                ./proxy/grpc
+COPY proxy/tower-router        ./proxy/tower-router
+COPY --from=libs /usr/src/conduit/target ./target
+ARG DEBUG
+RUN if [ -n "$DEBUG" ]; \
+    then cargo build -p conduit-grpc --features=arbitrary --frozen ; \
+    else cargo build -p conduit-grpc --features=arbitrary --frozen --release ; \
+    fi
+#RUN if [ -n "$SKIP_TESTS" ]; \
+#    then echo "tests skipped" ; \
+#    elif [ -n "$DEBUG" ]; \
+#    then cargo test -p conduit-grpc --features=arbitrary --frozen ; \
+#    else cargo test -p conduit-grpc --features=arbitrary --frozen --release ; \
+#    fi
+
+## Build the proxy binary using pre-built dependencies.
+FROM $RUST_IMAGE as bin
+WORKDIR /usr/src/conduit
+COPY --from=fetch $CARGO_HOME $CARGO_HOME
+COPY Cargo.toml Cargo.lock ./
 COPY proto ./proto
 COPY proxy ./proxy
-ARG RELEASE
-RUN if [ -z "$RELEASE" ]; \
-    then cargo build --frozen -p conduit-proxy           && mv target/debug/conduit-proxy   target/conduit-proxy ; \
-    else cargo build --frozen -p conduit-proxy --release && mv target/release/conduit-proxy target/conduit-proxy ; \
+COPY --from=grpc /usr/src/conduit/target ./target
+ARG DEBUG
+RUN if [ -n "$DEBUG" ]; \
+    then cargo build -p conduit-proxy --bin conduit-proxy --frozen ; \
+    else cargo build -p conduit-proxy --bin conduit-proxy --frozen --release ; \
+    fi
+ARG SKIP_TESTS
+RUN if [ -n "$SKIP_TESTS" ]; \
+    then echo "tests skipped" ; \
+    elif [ -n "$DEBUG" ]; \
```

Comment:
not quite:

1. if SKIP_TESTS is set, just echo,
2. otherwise, if debug is set, test the debug build
3. otherwise, test the release build

I really think that tests should be run by default when producing artifacts.